### PR TITLE
Check that Portal UI form outputs map to template parameter inputs on pull requests

### PR DESCRIPTION
**Pull request by glennmusa**
# Description

So that:

- On a PR, check to see that the Portal UI is compatible with the compiled Bicep deployment template
- check that all required parameters in the deployment template have a matching named output in the Portal UI template
- check that the Portal UI template outputs contain only named values that are available parameters in the deployment template

Here's what it looks like when this check fails:

https://ag-ascii.visualstudio.com/Mission%20Landing%20Zone%20-%20Pipeline/_build/results?buildId=6746&view=results

Here's what it looks like when this check passes:

https://ag-ascii.visualstudio.com/Mission%20Landing%20Zone%20-%20Pipeline/_build/results?buildId=6756&view=results

These proposed changes:

- adds an Azure DevOps Pipeline that executes on pull requests to `/main`
- when the pipeline executes, the `mlz.portal.json` declarative form and `mlz.json` deployment template are validated and if any errors occur, the pipeline fails
- references an external utility to perform the validation from [glennmusa/validate-declarative-form](https://github.com/glennmusa/validate-declarative-form)

## Issue reference

The issue this PR will close: #598

## Checklist

Please make sure you've completed the relevant tasks for this PR out of the following list:

* [x] All acceptance criteria in the backlog item are met
* [x] The documentation is updated to cover any new or changed features
* [x] Manual tests have passed
* [x] Relevant issues are linked to this PR

### CODE REVIEW

**Review by brooke-hamilton — CHANGES_REQUESTED** [review]
This PR is 🚀 🦖 🎉 . The README was easy to follow and I was able to reproduce everything documented there, including testing mismatched form and parameters. I had one issue with building the dev container, but I added a suggestion to fix it. 

If you think this is a general purpose tool, in the future it could move to https://github.com/glennmusa and we could reference it there.

**Comment by glennmusa** [comment]
> This PR is 🚀 🦖 🎉 . The README was easy to follow and I was able to reproduce everything documented there, including testing mismatched form and parameters. I had one issue with building the dev container, but I added a suggestion to fix it.
> 
> If you think this is a general purpose tool, in the future it could move to https://github.com/glennmusa and we could reference it there.

I like your idea to make use of this as a general purpose tool from a different repository:

1. this tool is not MLZ-specific
1. this tool should be superseded in time by an official product 

I'll update this branch so that it makes use of the tool in a pipeline, but host that tool elsewhere. Thanks for the idea.

File: .vscode/extensions.json [hunk 1]
```diff
@@ -3,16 +3,19 @@
...
+		"hashicorp.terraform",
 		"ms-vscode-remote.vscode-remote-extensionpack",
 		"davidanson.vscode-markdownlint",
 		"github.vscode-pull-request-github",
+		"timonwong.shellcheck",
 		"ms-azuretools.vscode-bicep",
+		"bierner.markdown-preview-github-styles",
+		"ms-azure-devops.azure-pipelines",
+		"ms-python.python",
+		"ms-python.vscode-pylance"
```
[the remark this answers is not in the document]
Comment (glennmusa) [reply .vscode/extensions.json]: Thanks for the thorough review, I've taken your suggestions here: https://github.com/glennmusa/validate-declarative-form/blob/530e329a7056518818388bccacd8097a7ed2c408/.devcontainer/devcontainer.json#L18-L43

File: src/bicep/form/validation/README.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+And you should get back some help:
+
+```plaintext
+usage: validate_declarative_form.py [-h] form_template_path deployment_template_path 
+
+Validate a Declarative Form UI template against an ARM Deployment Template.
+```
+
+So, execute it like:
```
[the remark this answers is not in the document]
Comment (glennmusa) [reply src/bicep/form/validation/README.md]: OBE -- moved this tool to its own repo: https://github.com/glennmusa/validate-declarative-form

File: src/bicep/form/validation/README.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+- You can run them all selecting the "Run Tests" button (or open the Command Pallete (F1) and type `Test: Run All Tests`)
+- You can debug individual tests by setting a breakpoint in `validate_declarative_form.py` or `validate_declarative_form_test.py` and selecting `Debug Test` from the tests pane
+
+### Testing from the terminal
+
+You can run the unit tests by calling the `unittest` module from Python like:
+
+```plaintext
+# cd src/bicep/form/validation
+python -m unittest validate_declarative_form_test.py -v
```
[the remark this answers is not in the document]
Comment (glennmusa) [reply src/bicep/form/validation/README.md]: Thanks for the thorough review, I've taken your suggestions here: https://github.com/glennmusa/validate-declarative-form/blob/530e329a7056518818388bccacd8097a7ed2c408/README.md?plain=1#L88-L91

File: src/bicep/form/validation/validate_declarative_form.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+        - json_file_path (string): the deployment template file path
+
+    Returns:
+        - json_as_object (object): the deployment template as a Python object
+    """
+
+    try:
+        with open(json_file_path, 'r', encoding="UTF-8") as json_file:
+            json_as_object = json.load(json_file)
+    except Exception as exception:
```
[the remark this answers is not in the document]
Comment (glennmusa) [reply src/bicep/form/validation/validate_declarative_form.py]: Thank you for the thorough review, I took your feedback and implemented it here: https://github.com/glennmusa/validate-declarative-form/blob/530e329a7056518818388bccacd8097a7ed2c408/validate_declarative_form.py#L126

File: .vscode/settings.json [hunk 5]
```diff
@@ -17,5 +17,16 @@
...
+    "python.linting.enabled": true,
+    "python.testing.unittestArgs": [
+        "-v",
+        "-s",
+        "./src/bicep/form/validation/",
+        "-p",
+        "*test.py"
+    ],
+    "python.testing.pytestEnabled": false,
+    "python.testing.unittestEnabled": true
```
[the remark this answers is not in the document]
Comment (glennmusa) [reply .vscode/settings.json]: This brings up another issue of housekeeping we can do within the repository. Push the team and contributors into using the devcontainer so we don't maintain these two separate and LIFO settings configurations.

**Comment by glennmusa** [comment]
Thanks for the thorough review @brooke-hamilton!

I've taken your advice and now host this tool here: https://github.com/glennmusa/validate-declarative-form

It's extremely lightweight today to `cURL` the single utility this change needs, but for completeness, I added an Issue in that backlog to expose the utility as a GitHub Action: https://github.com/glennmusa/validate-declarative-form/issues/4

**Review by brooke-hamilton — APPROVED** [review]

**Pull request merged**